Implement backend-agnostic rpc._wait_all_workers() utility

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31893 Delete all user forks tracked in RRefContext before shutting down
* **#32190 Implement backend-agnostic rpc._wait_all_workers() utility**

We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19399908](https://our.internmc.facebook.com/intern/diff/D19399908/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19399908/)!

xush6528 committed:
Implement backend-agnostic rpc._wait_all_workers() utility"

We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19399908](https://our.internmc.facebook.com/intern/diff/D19399908/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19399908/)!

[ghstack-poisoned]

xush6528 commented on torch/distributed/rpc/api.py:
```diff
@@ -120,10 +206,14 @@ def shutdown(graceful=True):
+    try:
+        # This might raise a `TORCH_CHECK()` exception.
         _destroy_rref_context(_ignore_rref_leak)
+    finally:
```
@mrshenli This fix the destruction segfault in Python3.5. We need to use finally clause to ensure running `_cleanup_python_rpc_handler()`

zhaojuanmao replied:
Thanks! would you please add comments for finally here so that we can understand it in the future

xush6528 replied:
@zhaojuanmao Updated the comment.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 2dddfafb:


* **1/3** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=4a26bb9b18d8c2cac7d23b6a2b021e9628701e99&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 4a26bb9b since Jan 14
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **2/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2dddfafb4faa31adc180b27fc8e69b40a288b678).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4244629) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86138069) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86138069))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
HEAD is now at 2dddfafb4f Implement backend-agnostic rpc._wait_all_workers() utility" 
+ git reset --hard 2dddfafb4faa31adc180b27fc8e69b40a288b678 
HEAD is now at 2dddfafb4f Implement backend-agnostic rpc._wait_all_workers() utility" 
+ git merge --allow-unrelated-histories --no-edit --no-ff 9bf0479b6564ab28229e1e0090e35648a811ea15 
Auto-merging torch/distributed/rpc/api.py 
CONFLICT (content): Merge conflict in torch/distributed/rpc/api.py 
Removing torch/csrc/distributed/rpc/script_functions.h 
Removing torch/csrc/distributed/rpc/script_functions.cpp 
Auto-merging test/rpc_test.py 
Auto-merging test/dist_autograd_test.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>

### 2 failures *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_cuda10_1_cudnn7_py3_nogpu_test</sup> | <sup>Test</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_cuda10_1_cudnn7_py3_NO_AVX_NO_AVX2_test</sup> | <sup>Test</sup> | &#128721; [Broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=4a26bb9b18d8c2cac7d23b6a2b021e9628701e99&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35&highlight_job=pytorch_linux_xenial_cuda10_1_cudnn7_py3_NO_AVX_NO_AVX2_test) |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 committed:
Update on "Implement backend-agnostic rpc._wait_all_workers() utility""


We need a backend-agnostic mechanism to do barrier-like operation before locally destroy RRef context and shutdown RPC Agent.

- Sort worker names.
- Elect the first name as the leader in the ordered worker names.
- Followers reports therir intent to synchronize to the leader.
- Leader also reports to itself, when `_wait_all_workers()` called.
- If all workers report their intent to proceed, leader send the command to every one to proceed.

Differential Revision: [D19399908](https://our.internmc.facebook.com/intern/diff/D19399908/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19399908/)!

[ghstack-poisoned]

mrshenli reviewed (approved):
Thanks for debugging this! This is awesome work!!
